lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3771,79 +3640,992 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+    auto insertion =
+        bufferMap.try_emplace({origBB, originalBuffer}, adjointBuffer);
+    assert(insertion.second); (void)insertion;
+  }
+
+  SILValue &getTangentBuffer(SILBasicBlock *origBB, SILValue originalBuffer) {
+    auto diffBuilder = getDifferentialBuilder();
+    assert(originalBuffer->getType().isAddress());
+    assert(originalBuffer->getFunction() == original);
+    auto insertion = bufferMap.try_emplace({origBB, originalBuffer},
+                                           SILValue());
+    if (!insertion.second) // not inserted
+      return insertion.first->getSecond();
+
+    // Set insertion point for local allocation builder: before the last local
+    // allocation, or at the start of the adjoint function's entry if no local
+    // allocations exist yet.
+    diffLocalAllocBuilder.setInsertionPoint(
+        getDifferential().getEntryBlock(),
+        getNextDifferentialLocalAllocationInsertionPoint());
+
+    // Allocate local buffer and initialize to zero.
+    auto bufObjectType = getRemappedTangentType(originalBuffer->getType());
+    auto *newBuf = diffLocalAllocBuilder.createAllocStack(
+        originalBuffer.getLoc(), bufObjectType);
+
+    // Temporarily change global builder insertion point and emit zero into the
+    // local buffer.
+    auto insertionPoint = diffLocalAllocBuilder.getInsertionBB();
+    diffBuilder.setInsertionPoint(
+        diffLocalAllocBuilder.getInsertionBB(),
+        diffLocalAllocBuilder.getInsertionPoint());
+    emitZeroIndirect(bufObjectType.getASTType(), newBuf, newBuf->getLoc());
+    diffBuilder.setInsertionPoint(insertionPoint);
+
+    // Create cleanup for local buffer.
+    differentialLocalAllocations.push_back(newBuf);
+    return (insertion.first->getSecond() = newBuf);
+  }
+
+  //--------------------------------------------------------------------------//
+  // Type transformer
+  //--------------------------------------------------------------------------//
+
+  Optional<VectorSpace> getTangentSpace(CanType type) {
+    return type->getAutoDiffAssociatedTangentSpace(
+        LookUpConformanceInModule(getModule().getSwiftModule()));
+  }
+
+  /// Assuming the given type conforms to `Differentiable` after remapping,
+  /// returns the associated tangent space type.
+  SILType getRemappedTangentType(SILType type) {
+    return SILType::getPrimitiveType(
+        getTangentSpace(remapType(type).getASTType())->getCanonicalType(),
+        type.getCategory());
+  }
+
+  //--------------------------------------------------------------------------//
+  // Tngent value mapping
+  //--------------------------------------------------------------------------//
+
+  /// Get the tangent for an original value. The given value must be in the
+  /// original function.
+  ///
+  /// This method first tries to find an entry in `tangentValueMap`. If a tangent
+  /// doesn't exist, create a zero tangent.
+  AdjointValue getTangentValue(SILValue originalValue) {
```

Comment:
For tangent values in forward mode, we should never run into a case where the tangent doesn't exist right? @rxwei @dan-zheng

Reply:
I believe you're right. You can try adding:
```cpp
  assert(tangentValueMap.count(originalValue));
  return tangentValueMap[originalValue];
```

Reply:
Awesome thanks!

Reply:
Not necessarily. What if you are differentiating `{ x in 0 }`?

Reply:
Yes goodpoint! I was thinking of how in the perfect cases this would work, but not in functions with a non varied result, like in the test case:
```swift
func one() -> Float {
  return 1
}
@differentiable
func nonVariedResult(_ x: Float) -> Float {
  // expected-warning @+1 2 {{result does not depend on differentiation arguments and will always have a zero derivative; do you want to use 'withoutDerivative(at:)'?}} {{10-10=withoutDerivative(at:}}
  return one()
}
```